From: Jiri Slaby <jslaby@suse.cz>
Date: Fri, 14 Oct 2016 07:37:32 +0200
Subject: fix get_builtin_firmware compilation with gcc 7
Patch-mainline: not yet, under discussion "firmware: declare __{start,end}_builtin_fw as pointers"
References: bnc#1004626

gcc 7 miscompiles the kernel since the kernel relies on undefined
behaviour. Fix this for get_builtin_firmware, so that the kernel boots
at least. The other VMLINUX_SYMBOLs have to be fixed too.

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 arch/x86/kernel/cpu/microcode/core.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/arch/x86/kernel/cpu/microcode/core.c b/arch/x86/kernel/cpu/microcode/core.c
index 5ce5155f0695..83f2bf7ff46d 100644
--- a/arch/x86/kernel/cpu/microcode/core.c
+++ b/arch/x86/kernel/cpu/microcode/core.c
@@ -97,9 +97,11 @@ extern struct builtin_fw __end_builtin_fw[];
 bool get_builtin_firmware(struct cpio_data *cd, const char *name)
 {
 #ifdef CONFIG_FW_LOADER
-	struct builtin_fw *b_fw;
+	struct builtin_fw *b_fw = __start_builtin_fw;
 
-	for (b_fw = __start_builtin_fw; b_fw != __end_builtin_fw; b_fw++) {
+	OPTIMIZER_HIDE_VAR(b_fw);
+
+	for (; b_fw != __end_builtin_fw; b_fw++) {
 		if (!strcmp(name, b_fw->name)) {
 			cd->size = b_fw->size;
 			cd->data = b_fw->data;
-- 
2.10.1

